### PR TITLE
Migrate to Dart 3.0

### PR DESCRIPTION
**Pull request by josxha**
#1512 (5fb3b3b222ba23d6ccd1261b395f045595ea502a) bumped the minimum Flutter SDK to 3.10, but did not change the minimum Dart SDK. This PR changes the Dart SDK to 3.0 minimum, enabling the new features and performance benefits.

In addition, this PR:

- Removes the Tuple dependency in favour of [Records](https://dart.dev/language/records)
- Uses new [pattern matching switches](https://dart.dev/language/patterns#switch-statements-and-expressions) in some locations
- Seals the `TileDisplay` class to benefit from new pattern matching switches, instead of switching on `runtimeType`
- Deprecates `TileUpdateTransformers.alwaysLoadAndPrune`, as it should be unnecessary compared to `ignoreTapEvents`
- Improves some documentation
- Includes some other more minor improvements/changes

This PR also fixes the failing Android builds by upgrading Gradle.

### CODE REVIEW

**Comment by JaffaKetchup** [comment]
Hi @josxha, and thanks for the contribution.
Personally, I can't bump the version to Dart 3, as I'm using FVM, and that is currently having some issues. This might hinder my ability to work on this project.
I'll check on a few things tomorrow. If everything turns out to be OK, I see no reason why this shouldn't be OK to merge (unless we decide we want to keep a little more compatibility).

**Comment by josxha** [comment]
Thanks for your message, @JaffaKetchup! I wasn't aware of the FVM project, but do hope that they manage to resolve the issues soon.

**Comment by JaffaKetchup** [comment]
For reference: https://github.com/fluttertools/fvm/issues/474.
When I try to build this project, my tools complain about the Dart SDK version being too high. Can't upgrade due to the above issue.
Will try to see if there's anything I can do to bypass this.

**Comment by JaffaKetchup** [comment]
@josxha It's pretty much that, but it also manages the global Flutter version. I've got my fingers crossed that it might be solved soon, the maintainer replied 10 hrs ago.
I've also got a really busy month coming up - but maybe if I get some time I'll remove FVM and just deal with one installation.

**Comment by JaffaKetchup** [comment]
I've managed to reinstall Dart 3, Flutter & FVM by using FVM from source instead of from Chocolatey.
I'll have a look at this soon hopefully, otherwise it'll be about 2-3 weeks.

File: lib/src/geo/crs/crs.dart
Comment (TesteurManiak) [inline lib/src/geo/crs/crs.dart:134]: Should be:
```dart
(double, double) get wrapLat => const (-85.06, 85.06);
```

File: lib/src/layer/marker_layer.dart
Comment (TesteurManiak) [inline lib/src/layer/marker_layer.dart:34]: You can use pattern matching for both statements:
```dart
if (pos.alignment case final align?) return Anchor._(width, height, align);
if (pos.anchor case final anchor?) return anchor;
```
This will avoid to use the `!` operator.

File: lib/src/layer/tile_layer/tile_bounds/tile_bounds.dart
Comment (TesteurManiak) [inline lib/src/layer/tile_layer/tile_bounds/tile_bounds.dart:127]: Same here, use pattern matching:
```dart
if (crs.wrapLng case final wrapLng?) {
  final wrapXMin = (crs.latLngToPoint(LatLng(0, wrapLng.$1), tzDouble).x / _tileSize).floor();

  // ...
}
```

File: lib/src/layer/tile_layer/tile_bounds/tile_bounds.dart
Comment (TesteurManiak) [inline lib/src/layer/tile_layer/tile_bounds/tile_bounds.dart:140]: Same here, use pattern matching:
```dart
if (crs.wrapLat case final wrapLat?) {
  final wrapYMin = (crs.latLngToPoint(LatLng(wrapLat.$1, 0), tzDouble).y / _tileSize).floor();

  // ...
}
```

File: lib/src/layer/tile_layer/tile_display.dart
Comment (TesteurManiak) [inline lib/src/layer/tile_layer/tile_display.dart:53]: Prefer this implementation as this will avoid the usage of the `as` keyword:
```dart
sealed class TileDisplay {
  // ...

  @optionalTypeArgs
  T? map<T extends Object?>({
    T? Function(InstantaneousTileDisplay)? instantaneous,
    T? Function(FadeInTileDisplay)? fadeIn,
  }) {
    final tileDisplay = this;
    return switch (tileDisplay) {
      InstantaneousTileDisplay() => instantaneous?.call(tileDisplay),
      FadeInTileDisplay() => fadeIn?.call(tileDisplay),
    };
  }
}
```

**Review by TesteurManiak — CHANGES_REQUESTED** [review]
Some minor changes to make before merging otherwise LGTM.

File: lib/src/layer/marker_layer.dart
Comment (JaffaKetchup) [reply lib/src/layer/marker_layer.dart:34]: (Was my change.)
Didn't realise this feature. Not too bad, but I think readability suffers a bit. Happy with it though.

```suggestion
    if (pos.alignment case final align?) return Anchor._(width, height, align);
    if (pos.anchor case final anchor?) return anchor;
```

File: lib/src/layer/tile_layer/tile_display.dart
Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_display.dart]: (Was my change.)
Resolved in 2b1b2f7, except for the (imo unnecessary) type narrowing on `T` and `@optionalTypeArgs` annotation. Also renamed it to `when` to better reflect what it does.

File: lib/src/layer/tile_layer/tile_bounds/tile_bounds.dart
Comment (josxha) [reply lib/src/layer/tile_layer/tile_bounds/tile_bounds.dart]: resolved in https://github.com/fleaflet/flutter_map/pull/1517/commits/3a58ee8354ff119a5168f94eef54aa1ac6a2f373

File: lib/src/layer/tile_layer/tile_bounds/tile_bounds.dart
Comment (josxha) [reply lib/src/layer/tile_layer/tile_bounds/tile_bounds.dart]: Resolved in https://github.com/fleaflet/flutter_map/pull/1517/commits/3a58ee8354ff119a5168f94eef54aa1ac6a2f373

File: lib/src/geo/crs/crs.dart
Comment (josxha) [reply lib/src/geo/crs/crs.dart]: https://github.com/fleaflet/flutter_map/pull/1517/commits/2bff4b7a32fbf9a41358e994449388de4ac25b19

**Review by TesteurManiak — APPROVED** [review]

**Comment by JaffaKetchup** [comment]
Thanks @josxha 🎉.

**Pull request merged**
